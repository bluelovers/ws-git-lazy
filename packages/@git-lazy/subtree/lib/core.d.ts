/**
 * Created by user on 2020/6/5.
 */
/// <reference types="node" />
import { EnumSubtreeCmd, IOptions, IOptionsRuntime } from './types';
export declare function handleOptions(options: IOptions): IOptionsRuntime;
export declare function _cmd(cmd: EnumSubtreeCmd, opts: IOptionsRuntime): import("cross-spawn-extra").SpawnASyncReturnsPromise<string | Buffer>;
export declare function _call(cmd: EnumSubtreeCmd, options: IOptions): import("cross-spawn-extra").SpawnASyncReturnsPromise<string | Buffer>;
