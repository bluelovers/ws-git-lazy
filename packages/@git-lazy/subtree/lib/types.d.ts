/**
 * Created by user on 2020/6/5.
 */
import { ITSRequireAtLeastOne } from 'ts-type';
export declare type IOptions = {
    prefix: string;
    cwd?: string;
    branch?: string;
    squash?: boolean;
} & ITSRequireAtLeastOne<{
    remote: string;
    name?: string;
}>;
export interface IOptionsRuntime {
    options: IOptions;
    cwd: string;
    root: string;
    remote: string;
    branch: string;
    prefixType: EnumPrefixType;
    prefix: string;
    prefixPath: string;
}
export declare enum EnumPrefixType {
    ROOT = 0,
    RELATIVE = 1,
    ABSOLUTE = 2
}
export declare enum EnumSubtreeCmd {
    add = "add",
    push = "push",
    pull = "pull"
}
