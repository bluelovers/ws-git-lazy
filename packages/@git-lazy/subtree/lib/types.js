"use strict";
/**
 * Created by user on 2020/6/5.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumSubtreeCmd = exports.EnumPrefixType = void 0;
var EnumPrefixType;
(function (EnumPrefixType) {
    EnumPrefixType[EnumPrefixType["ROOT"] = 0] = "ROOT";
    EnumPrefixType[EnumPrefixType["RELATIVE"] = 1] = "RELATIVE";
    EnumPrefixType[EnumPrefixType["ABSOLUTE"] = 2] = "ABSOLUTE";
})(EnumPrefixType = exports.EnumPrefixType || (exports.EnumPrefixType = {}));
var EnumSubtreeCmd;
(function (EnumSubtreeCmd) {
    EnumSubtreeCmd["add"] = "add";
    EnumSubtreeCmd["push"] = "push";
    EnumSubtreeCmd["pull"] = "pull";
    EnumSubtreeCmd["split"] = "split";
})(EnumSubtreeCmd = exports.EnumSubtreeCmd || (exports.EnumSubtreeCmd = {}));
//# sourceMappingURL=types.js.map