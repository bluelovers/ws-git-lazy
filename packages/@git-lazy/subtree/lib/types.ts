/**
 * Created by user on 2020/6/5.
 */

import { ITSRequireAtLeastOne } from 'ts-type';

export type IOptions = {

	prefix: string,

	cwd?: string,
	branch?: string,

	squash?: boolean,

} & ITSRequireAtLeastOne<{
	remote: string,
	name?: string,
}>

export interface IOptionsRuntime
{
	options: IOptions;

	cwd: string;
	root: string;

	remote: string;
	branch: string;

	prefixType: EnumPrefixType;
	prefix: string;
	prefixPath: string;
}

export enum EnumPrefixType
{
	ROOT,
	RELATIVE,
	ABSOLUTE,
}

export enum EnumSubtreeCmd
{
	add = 'add',
	push = 'push',
	pull = 'pull',
}
