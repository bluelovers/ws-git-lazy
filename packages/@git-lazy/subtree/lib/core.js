"use strict";
/**
 * Created by user on 2020/6/5.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._call = exports._cmd = exports.handleOptions = void 0;
const types_1 = require("./types");
const upath2_1 = require("upath2");
const git_root2_1 = __importDefault(require("git-root2"));
const util_1 = require("./util");
const spawn_1 = __importDefault(require("@git-lazy/spawn"));
function handleOptions(options) {
    var _a, _b, _c;
    let cwd = upath2_1.resolve((_a = options.cwd) !== null && _a !== void 0 ? _a : process.cwd());
    let root = upath2_1.normalize(git_root2_1.default(cwd));
    let { prefix, prefixType } = util_1.handlePrefix(options.prefix);
    let prefixPath = prefix;
    if (prefixType !== types_1.EnumPrefixType.ROOT) {
        prefixPath = upath2_1.resolve(cwd, prefix);
        if (util_1.inSubPath(prefixPath, root)) {
            prefixPath = upath2_1.relative(root, prefixPath);
        }
        else {
            throw new Error(`prefix path is not allow: ${prefixPath}`);
        }
    }
    let branch = (_b = options.branch) !== null && _b !== void 0 ? _b : 'master';
    let remote = (_c = options.remote) !== null && _c !== void 0 ? _c : options.name;
    if (typeof remote !== 'string' || !remote.length) {
        throw new TypeError(`remote is not valid: ${remote}`);
    }
    if (typeof branch !== 'string' || !branch.length) {
        throw new TypeError(`branch is not valid: ${branch}`);
    }
    return {
        options,
        cwd,
        root,
        remote,
        branch,
        prefixType,
        prefix,
        prefixPath,
    };
}
exports.handleOptions = handleOptions;
function _cmd(cmd, opts) {
    return spawn_1.default('git', [
        'subtree',
        cmd,
        opts.remote,
        opts.branch,
        '--prefix',
        opts.prefixPath,
        ...(opts.options.squash ? ['--squash'] : []),
    ], {
        cwd: opts.root,
        stdio: 'inherit',
    });
}
exports._cmd = _cmd;
function _call(cmd, options) {
    let opts = handleOptions(options);
    return _cmd(cmd, opts);
}
exports._call = _call;
//# sourceMappingURL=core.js.map