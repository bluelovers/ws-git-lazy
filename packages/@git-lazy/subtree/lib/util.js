"use strict";
/**
 * Created by user on 2020/6/5.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.inSubPath = exports.handlePrefix = void 0;
const upath2_1 = require("upath2");
const path_1 = require("path");
const types_1 = require("./types");
function handlePrefix(prefix) {
    const bool = /^\.[\/\\]/.test(prefix);
    prefix = upath2_1.normalize(prefix);
    if (bool && !prefix.startsWith('../')) {
        prefix = './' + prefix;
    }
    let prefixType = types_1.EnumPrefixType.ROOT;
    if (/^\/[^\/]/.test(prefix)) {
        prefix = prefix.slice(1);
    }
    else if (prefix.startsWith('../') || prefix.startsWith('./')) {
        prefixType = types_1.EnumPrefixType.RELATIVE;
    }
    else if (path_1.isAbsolute(prefix)) {
        prefixType = types_1.EnumPrefixType.ABSOLUTE;
    }
    return {
        prefixType,
        prefix,
    };
}
exports.handlePrefix = handlePrefix;
function inSubPath(sub, root) {
    let r = upath2_1.normalize(root);
    let s = upath2_1.normalize(sub);
    return s.indexOf(r) === 0 && s.length > r.length;
}
exports.inSubPath = inSubPath;
//# sourceMappingURL=util.js.map