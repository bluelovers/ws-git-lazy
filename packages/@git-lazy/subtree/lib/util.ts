/**
 * Created by user on 2020/6/5.
 */

import { normalize } from 'upath2';
import { isAbsolute } from "path";
import { EnumPrefixType } from './types';

export function handlePrefix(prefix: string)
{
	const bool = /^\.[\/\\]/.test(prefix);

	prefix = normalize(prefix);

	if (bool && !prefix.startsWith('../'))
	{
		prefix = './' + prefix;
	}

	let prefixType: EnumPrefixType = EnumPrefixType.ROOT;

	if (/^\/[^\/]/.test(prefix))
	{
		prefix = prefix.slice(1);
	}
	else if (prefix.startsWith('../') || prefix.startsWith('./'))
	{
		prefixType = EnumPrefixType.RELATIVE;
	}
	else if (isAbsolute(prefix))
	{
		prefixType = EnumPrefixType.ABSOLUTE;
	}

	return {
		prefixType,
		prefix,
	}
}

export function inSubPath(sub: string, root: string)
{
	let r = normalize(root)
	let s = normalize(sub)

	return s.indexOf(r) === 0 && s.length > r.length
}
