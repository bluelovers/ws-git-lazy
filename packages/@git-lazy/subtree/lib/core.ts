/**
 * Created by user on 2020/6/5.
 */

import { EnumSubtreeCmd, IOptions, IOptionsRuntime, EnumPrefixType } from './types';
import { resolve, normalize, relative } from 'upath2';
import gitRoot from 'git-root2';
import { handlePrefix, inSubPath } from './util';
import crossSpawnGitAsync from '@git-lazy/spawn';
import debug from '@git-lazy/debug';

export function handleOptions(options: IOptions): IOptionsRuntime
{
	let cwd = resolve(options.cwd ?? process.cwd());
	let root = normalize(gitRoot(cwd));

	let { prefix, prefixType } = handlePrefix(options.prefix);
	let prefixPath = prefix;

	if (prefixType !== EnumPrefixType.ROOT)
	{
		prefixPath = resolve(cwd, prefix)

		if (inSubPath(prefixPath, root))
		{
			prefixPath = relative(root, prefixPath)
		}
		else
		{
			throw new Error(`prefix path is not allow: ${prefixPath}`)
		}
	}

	let branch = options.branch ?? 'master';
	let remote = options.remote ?? options.name;

	if (typeof remote !== 'string' || !remote.length)
	{
		throw new TypeError(`remote is not valid: ${remote}`)
	}

	if (typeof branch !== 'string' || !branch.length)
	{
		throw new TypeError(`branch is not valid: ${branch}`)
	}

	return {
		options,
		cwd,
		root,

		remote,
		branch,

		prefixType,
		prefix,
		prefixPath,
	}
}

export function _cmd(cmd: EnumSubtreeCmd, opts: IOptionsRuntime)
{
	return crossSpawnGitAsync('git', [
		'subtree',
		cmd,
		opts.remote,
		opts.branch,
		'--prefix',
		opts.prefixPath,
		...(opts.options.squash ? ['--squash'] : []),
	], {
		cwd: opts.root,
		stdio: 'inherit',
	})
}

export function _call(cmd: EnumSubtreeCmd, options: IOptions)
{
	let opts = handleOptions(options)

	return _cmd(cmd, opts)
}
