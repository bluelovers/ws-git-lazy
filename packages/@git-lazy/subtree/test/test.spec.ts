import { handlePrefix } from '../lib/util';

describe(`prefix`, () =>
{

	[
		`./node-novel-info`,
		`.\\node-novel-info`,

		`../node-novel-info`,
		`..\\node-novel-info`,

		`packages/node-novel-info`,
		`packages\\node-novel-info`,


		`//node-novel-info`,
		`\\\\node-novel-info`,

		`/node-novel-info`,
		`\\node-novel-info`,
	]
		.forEach(prefix =>
		{
			test(prefix, () =>
			{

				let actual = handlePrefix(prefix);
				let expected;

				expect(actual).toMatchSnapshot();

			});
		})
	;

})
