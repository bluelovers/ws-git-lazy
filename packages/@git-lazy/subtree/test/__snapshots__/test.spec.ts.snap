// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prefix ../node-novel-info 1`] = `
Object {
  "prefix": "../node-novel-info",
  "prefixType": 1,
}
`;

exports[`prefix ..\\node-novel-info 1`] = `
Object {
  "prefix": "../node-novel-info",
  "prefixType": 1,
}
`;

exports[`prefix ./node-novel-info 1`] = `
Object {
  "prefix": "./node-novel-info",
  "prefixType": 1,
}
`;

exports[`prefix .\\node-novel-info 1`] = `
Object {
  "prefix": "./node-novel-info",
  "prefixType": 1,
}
`;

exports[`prefix //node-novel-info 1`] = `
Object {
  "prefix": "node-novel-info",
  "prefixType": 0,
}
`;

exports[`prefix /node-novel-info 1`] = `
Object {
  "prefix": "node-novel-info",
  "prefixType": 0,
}
`;

exports[`prefix \\\\node-novel-info 1`] = `
Object {
  "prefix": "node-novel-info",
  "prefixType": 0,
}
`;

exports[`prefix \\node-novel-info 1`] = `
Object {
  "prefix": "node-novel-info",
  "prefixType": 0,
}
`;

exports[`prefix packages/node-novel-info 1`] = `
Object {
  "prefix": "packages/node-novel-info",
  "prefixType": 0,
}
`;

exports[`prefix packages\\node-novel-info 1`] = `
Object {
  "prefix": "packages/node-novel-info",
  "prefixType": 0,
}
`;
